File: templates/nodejs-service/service/src/transport/rmqRpcServer.ts
import amqplib from "amqplib";
import { RPCServer, RPCHandler } from "./interfaces";
import { EnvelopedMessage } from "./envelopedMessage";
import { rmqDeclareAndBindQueue, rmqConsume } from "./rmqHelper";
import { rmqRpcReplyOutput, rmqRpcReplyError } from "./rmqRpcHelper";
import { RmqEventMap } from "./rmqEventMap";

export class RmqRPCServer implements RPCServer {
    private handlers: { [functionName: string]: RPCHandler };

    constructor(private logger: Console, private connection: amqplib.Connection, private eventMap: RmqEventMap) {
        this.handlers = {};
    }

    public registerHandler(functionName: string, handler: RPCHandler): RPCServer {
        this.handlers[functionName] = handler;
        return this;
    }

    public serve(): Promise<void> {
        const self = this;
        return new Promise(async (_, reject) => {
            try {
                const ch = await self.connection.createChannel();
                self.connection.on("error", (err) => {
                    reject(err);
                });
                self.connection.on("close", (err) => {
                    reject(err);
                })
                this.pServe(ch);
            } catch (err) {
                reject(err);
            }
        });
    }

    private async declareDlxAndBuildArgs(ch: amqplib.Channel, functionName: string): Promise<amqplib.Options.AssertQueue> {
        const args: amqplib.Options.AssertQueue = {};
        if (this.eventMap.getTtl(functionName) > 0) {
            const deadLetterExchange = this.eventMap.getDeadLetterExchange(functionName);
            const deadLetterQueue = this.eventMap.getDeadLetterQueue(functionName);
            await rmqDeclareAndBindQueue(ch, deadLetterExchange, deadLetterQueue, { durable: true });
            args.deadLetterExchange = deadLetterExchange;
            args.messageTtl = this.eventMap.getTtl(functionName);
        }
        return args;
    }

    private async pServe(ch: amqplib.Channel) {
        for (let key in this.handlers) {
            const functionName = key;
            // declare dlx
            const args = await this.declareDlxAndBuildArgs(ch, functionName);
            // declare queue
            const exchangeName = this.eventMap.getExchangeName(functionName);
            const queueName = this.eventMap.getExchangeName(functionName);
            const handler = this.handlers[functionName];
            await rmqDeclareAndBindQueue(ch, exchangeName, queueName, args);
            this.logger.log(`[INFO RmqRPCServer] Serve ${functionName}`);
            // consume
            const autoAck = this.eventMap.getAutoAck(functionName);
            rmqConsume(ch, queueName, args, autoAck, async (rmqMessageOrNull) => {
                try {
                    const rmqMessage = rmqMessageOrNull as amqplib.ConsumeMessage;
                    if (!autoAck) {
                        ch.ack(rmqMessage);
                    }
                    const { replyTo } = rmqMessage.properties;
                    const jsonEnvelopedInput = rmqMessage.content.toString();
                    const envelopedInput = new EnvelopedMessage(jsonEnvelopedInput);
                    const inputs = envelopedInput.getMessage().inputs;
                    try {
                        const output = await handler(...inputs);
                        this.logger.log(`[INFO RmqRPCServer] Reply ${functionName} by Sending to Queue ${replyTo}: `, JSON.stringify(output));
                        try {
                            await rmqRpcReplyOutput(ch, replyTo, envelopedInput, output);
                        } catch (err) {
                            this.logger.log(`[INFO RmqRPCServer] Failed to Reply ${functionName} by Sending to Queue ${replyTo}: `, err);
                        }
                    } catch (err) {
                        this.logger.error(`[ERROR RmqRPCServer] Reply Error ${functionName} by Sending to queue ${replyTo}: `, err);
                        await rmqRpcReplyError(ch, replyTo, envelopedInput, err);
                    }
                } catch (err) {
                    this.logger.error(`[ERROR RmqRPCServer] Error Replying ${functionName}: `, err);
                }
            });
        }
    }

}

File: templates/go-service/service/transport/rmqRpcServer.go
package transport

import (
	"fmt"
	"log"

	"github.com/streadway/amqp"
)

// CreateRmqRPCServer create new RmqRPC
func CreateRmqRPCServer(logger *log.Logger, connection *amqp.Connection, eventMap *RmqEventMap) *RmqRPCServer {
	return &RmqRPCServer{
		connection: connection,
		handlers:   map[string]RPCHandler{},
		logger:     logger,
		eventMap:   eventMap,
	}
}

// RmqRPCServer implementation
type RmqRPCServer struct {
	connection *amqp.Connection
	handlers   map[string]RPCHandler
	logger     *log.Logger
	eventMap   *RmqEventMap
}

// RegisterHandler register servicemap for call
func (s *RmqRPCServer) RegisterHandler(functionName string, handler RPCHandler) RPCServer {
	s.handlers[functionName] = handler
	return s
}

// Serve serve RPC
func (s *RmqRPCServer) Serve(errChan chan error) {
	// create connection and channel
	ch, err := s.connection.Channel()
	if err != nil {
		s.handleRmqError(err, errChan)
		return
	}
	amqpErrChan := s.connection.NotifyClose(make(chan *amqp.Error))
	defer ch.Close()
	for functionName, handler := range s.handlers {
		// declare dlx
		var args amqp.Table = nil
		if s.eventMap.GetTTL(functionName) > 0 {
			deadLetterExchange := s.eventMap.GetDeadLetterExchange(functionName)
			deadLetterQueue := s.eventMap.GetDeadLetterQueue(functionName)
			if _, err = rmqDeclareAndBindQueue(ch, deadLetterExchange, deadLetterQueue, true, false, false, false, nil); err != nil {
				s.handleRmqError(err, errChan)
				return
			}
			args = s.eventMap.GetQueueArgs(functionName)
		}
		// declare queue
		exchangeName := s.eventMap.GetExchangeName(functionName)
		queueName := s.eventMap.GetQueueName(functionName)
		_, err = rmqDeclareAndBindQueue(ch, exchangeName, queueName, true, false, false, false, args)
		if err != nil {
			s.handleRmqError(err, errChan)
			return
		}
		// start consume
		s.logger.Printf("[INFO RmqRPCServer] Serve %s", functionName)
		autoAck := s.eventMap.GetAutoAck(functionName)
		rmqMessages, err := rmqConsume(ch, queueName, true, false, false, false, autoAck, args)
		if err != nil {
			s.handleRmqError(err, errChan)
			return
		}
		// handle message
		thisHandler, thisFunctionName := handler, functionName
		go s.handleRmqMessages(ch, thisFunctionName, thisHandler, rmqMessages, autoAck)
	}
	err = <-amqpErrChan
	errChan <- err
}

func (s *RmqRPCServer) handleRmqMessages(ch *amqp.Channel, functionName string, handler RPCHandler, rmqMessages <-chan amqp.Delivery, autoAck bool) {
	for rmqMessage := range rmqMessages {
		replyTo := rmqMessage.ReplyTo
		envelopedInput, err := CreateEnvelopedMessageFromJSON(rmqMessage.Body)
		if err != nil {
			s.handleRmqMessageError(ch, functionName, replyTo, envelopedInput, err)
			if !autoAck {
				rmqMessage.Nack(false, true)
			}
			continue
		}
		inputs, err := envelopedInput.Message.GetInterfaceArray("inputs")
		if err != nil {
			s.handleRmqMessageError(ch, functionName, replyTo, envelopedInput, err)
			if !autoAck {
				rmqMessage.Nack(false, true)
			}
			continue
		}
		output, err := handler(inputs...)
		if err != nil {
			s.handleRmqMessageError(ch, functionName, replyTo, envelopedInput, err)
			if !autoAck {
				rmqMessage.Nack(false, true)
			}
			continue
		} else if !autoAck {
			rmqMessage.Ack(false)
		}
		s.logger.Printf("[INFO RmqRPCServer] Reply %s by Sending to Queue %s: %#v", functionName, replyTo, output)
		if err = rmqRPCReplyOutput(ch, replyTo, envelopedInput, output); err != nil {
			s.logger.Printf("[INFO RmqRPCServer] Failed to Reply %s by Sending to Queue %s: %#v", functionName, replyTo, err)
		}
	}
}

func (s *RmqRPCServer) handleRmqError(err error, errChan chan error) {
	s.logger.Println("[ERROR RmqRPCServer]", err)
	errChan <- err
}

func (s *RmqRPCServer) handleRmqMessageError(ch *amqp.Channel, thisFunctionName, replyTo string, envelopedInput *EnvelopedMessage, err error) {
	s.logger.Printf(fmt.Sprintf("[ERROR RmqRPCServer] Reply Error %s by Sending to Queue %s:", thisFunctionName, replyTo), err)
	if err = rmqRPCReplyError(ch, replyTo, envelopedInput, err); err != nil {
		s.logger.Printf("[ERROR RmqRPCServer] Failed to Reply Error %s by Sending to Queue %s: %#v", thisFunctionName, replyTo, err)
	}
}
